File: HackerRank/30DaysOfCode/CPP/day_18_stack_queue.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
    private:
        stack<char> st;
        queue<char> q;
    public:
        // plate store (LIFO)
        void pushCharacter(char c){
            st.push(c);
        }
        // train ticket line... (FIFO)
        void enqueueCharacter(char c){
            q.push(c);
        }

        char popCharacter(){
            char top = st.top();
            st.pop();
            return top;
        }

        char dequeueCharacter(){
            char front = q.front();
            q.pop();
            return front;
        }
};


int main(){
    string s;
    getline(cin, s);

    Solution sObj;

    for(int i = 0; i < s.length(); i++){
        sObj.pushCharacter(s[i]);
        sObj.enqueueCharacter(s[i]);
    }

    bool isPalindrome = true;

    for(int i = 0; i < s.length()/2; i++){
        cout << sObj.popCharacter() << "-" << sObj.dequeueCharacter() << endl;
        if(sObj.popCharacter() != sObj.dequeueCharacter()){
            isPalindrome = false;
            break;
        }
    }

    if (isPalindrome){
        cout << "The word, " << s << ", is a palindrome." << endl;
    } else {
        cout << "The word, " << s << ", is not a palindrome." << endl;
    }
    return 0;
}

File: HackerRank/30DaysOfCode/Python/day_18_stack_queue.py
class Solution:
    def __init__(self):
        self.stack = []
        self.queue = []

    def pushCharacter(self, ch):
        self.stack.append(ch)
    
    def enqueueCharacter(self, ch):
        self.queue.append(ch)
    
    def popCharacter(self):
        return self.stack.pop()
    
    def dequeueCharacter(self):
        return self.queue.pop(0)


if __name__ == '__main__':
    inp = input()
    s = Solution()
    for i in inp:
        s.pushCharacter(i)
        s.enqueueCharacter(i)
    
    isPalindrome = True
    for i in range(len(inp) // 2):
        if s.popCharacter() != s.dequeueCharacter():
            isPalindrome = False
            break
    if isPalindrome:
        print("The word, "+ inp +", is a palindrome.")
    else:
        print("The word, "+ inp +", is not a palindrome.")